src/platform/backends/libvirt/libvirt_wrapper.cpp

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright (C) 2019 Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "libvirt_wrapper.h"
+
+#include <multipass/logging/log.h>
+
+#include <dlfcn.h>
+
+namespace mp = multipass;
+namespace mpl = multipass::logging;
+
+namespace
+{
+constexpr auto logging_category = "libvirt-wrapper";
+
+auto open_libvirt_handle(const std::string& filename)
+{
+    // If filename is empty, it's for testing, ie, opening the test executable itself
+    auto handle = dlopen(filename.empty() ? nullptr : filename.c_str(), RTLD_NOW | RTLD_GLOBAL);
+
+    if (!handle)
+        throw mp::LibvirtOpenException(dlerror());
+
+    return handle;
+}
+
+auto get_symbol_address_for(const std::string& symbol, void* handle)
+{
+    auto address = dlsym(handle, symbol.c_str());
+
+    if (!address)
+        throw mp::LibvirtSymbolAddressException(symbol, dlerror());
+
+    return address;
+}
+} // namespace
+
+mp::LibvirtWrapper::LibvirtWrapper(const std::string& filename) : filename{filename}
+{
+    try
+    {
+        initialize_libvirt_functions();
+    }
+    catch (const std::exception& e)
+    {
+        mpl::log(mpl::Level::error, logging_category, e.what());
```

Comment:
No, I don't want the exception to bubble up in this case since I want the parent ctors to finish construction and not fail.  This way, if someone loads `libvirt` *after* `multipassd` starts up, `multipassd` can detect this during the health check and be happy :grin:

Reply:
I am an adept of always leaving constructed objects in a fully functional state, that is what constructors are for. Finishing construction only to require the client to check the state all the time, calling an init function if needed, is c-style and goes against separation of responsibilities.

So, why not throw from the constructor when it can't succeed, leaving it to the client  (the factory) to decide how to handle that? If what it wants to do is check and retry creating it, a (unique) pointer is the modular construct to achieve that IMO. That would do away with the `is_enabled`/`init` stuff.

A separate point is that the `catch` should only filter the specific exceptions it is handling if not re-throwing.

Reply:
Also, I tried to provoke an issue and after the error:

```
[2019-11-13T18:25:04.710] [error] [libvirt-wrapper] Failed to open the libvirt object: BADvirt.so: cannot open shared object file: No such file or directory
```

It seems the daemon does not reply to the client. I just went back to that client shell and the spinner was still going...

Reply:
~~[...] spinner still going [...]~~

Please ignore that, I had it stopped by gdb...

Reply:
Ok, thanks for the feedback.  I generally agree with what you are saying and will think about that wrt to this and see if I can come up with an RAII way of doing this.

Reply:
Ok, I think I have come up with a much better approach to this.  Let me know what you think!

Thanks!